File: Combat/Ring.cs
using CombatCavallers.cops;
using CombatCavallers.Lluitador;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Linq;

namespace CombatCavallers.Combat
{
    public class Ring : IRing
    {
        private const int VidaInicialDelsCombatents = 20;

        private readonly Random _rnd;
        private readonly ICombatents[] _Lluitadors;
        private int[] copsIlegals;
        private readonly ILogger<Ring> _logger;

        public Ring(ILogger<Ring> logger)
        {
            _rnd = new Random();
            _Lluitadors = new Resultat[2];
            copsIlegals = new int[2];
            _logger = logger;
        }

        /// <summary>
        /// Col·loca dos lluitadors al ring
        /// </summary>
        /// <param name="lluitador1">Primer lluitador</param>
        /// <param name="lluitador2">Segon lluitador</param>
        public void EntradaLluitadors(ILluitador lluitador1, ILluitador lluitador2)
        {
            _Lluitadors[0] = new Resultat(lluitador1, VidaInicialDelsCombatents);
            _Lluitadors[1] = new Resultat(lluitador2, VidaInicialDelsCombatents);
            copsIlegals = new int[2];
        }

        /// <summary>
        /// Fa que els dos lluitadors lluitin fins que s'acabi el combat
        /// </summary>
        /// <returns>Retorna una llista IResultat amb el resultat del combat</returns>
        public IEnumerable<IResultat> Lluiteu()
        {
            if (_Lluitadors.Any(l => l == null))
            {
                _logger.LogError("No es pot començar el combat. Falten Lluitadors");
                return null;
            }

            var elQuePica = _rnd.Next(2);
            _logger.LogInformation($"Sorteig de qui comença: .... {_Lluitadors[elQuePica].Nom}");

            while (_Lluitadors.All(l => l.EsKo() == false && l.EstaEliminat() == false))
            {
                var elQueRep = (elQuePica + 1) % 2;
                var proteccio = _Lluitadors[elQueRep].Lluitador.Protegeix();
                var (_, pica) = _Lluitadors[elQuePica].Lluitador.Pica();

                var efecteSobreDefensor = 0;
                var efecteSobreAtacant = 0;
                var missatge = "";

                switch (pica)
                {
                    case LlocOnPicar.Cap:
                    case LlocOnPicar.CostatDret:
                    case LlocOnPicar.CostatEsquerra:
                    case LlocOnPicar.Panxa:
                        efecteSobreDefensor = 1;
                        efecteSobreAtacant = 0;
                        missatge = $"{_Lluitadors[elQueRep].Nom} ({_Lluitadors[elQueRep].Vida}) rep un cop al {pica} de {_Lluitadors[elQuePica].Nom} ({_Lluitadors[elQuePica].Vida})";
                        break;
                    case LlocOnPicar.CopIlegal:
                        efecteSobreDefensor = 1;
                        efecteSobreAtacant = 0;
                        missatge = $"{_Lluitadors[elQueRep].Nom} ({_Lluitadors[elQueRep].Vida}) rep un cop ILEGAL de {_Lluitadors[elQuePica].Nom} ({_Lluitadors[elQuePica].Vida})";
                        break;
                    default:
                        missatge = "No ha passat res";
                        break;
                }

                var haRebut = proteccio.Any(l => l == pica) || pica == LlocOnPicar.CopIlegal;
                if (haRebut)
                {
                    _Lluitadors[elQueRep].TreuVida(efecteSobreDefensor);
                    _Lluitadors[elQuePica].TreuVida(efecteSobreAtacant);
                    _logger.LogInformation(missatge);
                }
                else
                {
                    _logger.LogInformation($"{_Lluitadors[elQueRep].Nom} atura el cop al {pica} de {_Lluitadors[elQuePica].Nom}");
                }

                if (pica is LlocOnPicar.CopIlegal)
                {
                    copsIlegals[elQuePica]++;
                    if (copsIlegals[elQuePica] == 3)
                    {
                        _logger.LogInformation($" {_Lluitadors[elQuePica].Nom} ELIMINAT PER COPS ILEGALS");
                        _Lluitadors[elQueRep].Elimina();
                    }
                }

                _logger.LogDebug($"{_Lluitadors[0].Nom}-({_Lluitadors[0].Vida}) vs {_Lluitadors[1].Nom}-({_Lluitadors[1].Vida})");
                elQuePica = elQueRep;
            }

            var guanyador = _Lluitadors.First(l => l.EsKo() == false && l.EstaEliminat() == false);
            var perdedor = _Lluitadors.First(l => l.EsKo() == true || l.EstaEliminat() == true);

            var comentariLocutor = "";

            if (perdedor.EstaEliminat())
            {
                comentariLocutor = $"{perdedor.Nom} està ELIMINAT!";
            }
            else
            {
                comentariLocutor = (guanyador.Vida - perdedor.Vida) > 5 ? "Quina Pallissa!!" : "";
            }

            _logger.LogInformation($"{perdedor.Nom} cau a terra!");
            _logger.LogInformation($"VICTÒRIA DE {guanyador.Nom}!!! {comentariLocutor}");

            return _Lluitadors;

        }
    }
}

File: combat/Ring.go
package combat

import (
	"errors"
	"fmt"
	"log"
	"math/rand"

	"github.com/utrescu/CombatCavallers.Go/cops"
	"github.com/utrescu/CombatCavallers.Go/lluitador"
)

// Ring és on es desenvolupen els combats
type Ring struct {
	resultat    []Resultat
	copsIlegals []int
}

// EntrenJugadors entren els jugadors al ring
func (ring *Ring) EntrenJugadors(jugador1 lluitador.ILluitador, jugador2 lluitador.ILluitador) {
	ring.resultat = make([]Resultat, 2)
	ring.resultat[0] = CreateResultat(jugador1, 20)
	ring.resultat[1] = CreateResultat(jugador2, 20)
	ring.copsIlegals = make([]int, 2)
	ring.copsIlegals[0] = 0
	ring.copsIlegals[1] = 0
}

// Lluiteu és la ordre de començar el combat
func (ring *Ring) Lluiteu() ([]Resultat, error) {
	if ring.resultat == nil {
		return nil, errors.New("ERROR Falten lluitadors")
	}

	elQuePica := rand.Intn(2)
	log.Printf("Sorteig de qui comença: .... %s", ring.resultat[elQuePica].GetNom())

	for ring.resultat[0].EsKo() == false && ring.resultat[1].EsKo() == false {
		elQueRep := (elQuePica + 1) % 2

		proteccio := ring.resultat[elQueRep].GetLluitador().Protegeix()
		_, pica := ring.resultat[elQuePica].GetLluitador().Pica()

		efecteSobreDefensor := 0
		efecteSobreAtacant := 0
		missatge := ""

		switch pica {
		case cops.Cap, cops.CostatDret, cops.CostatEsquerra, cops.Panxa:
			efecteSobreDefensor = 1
			efecteSobreAtacant = 0
			missatge = fmt.Sprintf("%s rep un cop %s de %s", ring.resultat[elQueRep].GetNom(), pica, ring.resultat[elQuePica].GetNom())
			break
		case cops.Collons:
			efecteSobreDefensor = 1
			efecteSobreAtacant = 0
			missatge = fmt.Sprintf("%s rep un cop ILEGAL de %s", ring.resultat[elQueRep].GetNom(), ring.resultat[elQuePica].GetNom())
			break
		default:
			missatge = "No ha passat res"
			break
		}

		haRebut := contains(proteccio, pica) || pica == cops.Collons

		if haRebut {
			ring.resultat[elQueRep].TreuVida(efecteSobreDefensor)
			ring.resultat[elQuePica].TreuVida(efecteSobreAtacant)
			log.Printf(missatge)
		} else {
			log.Printf("%s atura el cop %s de %s", ring.resultat[elQueRep].GetNom(), pica, ring.resultat[elQuePica].GetNom())
		}

		if pica == cops.Collons {
			ring.copsIlegals[elQuePica]++
			if ring.copsIlegals[elQuePica] == 3 {
				ring.resultat[elQuePica].Elimina()
				break
			}
		}

		log.Printf("%s-(%d) vs %s-(%d)", ring.resultat[0].GetNom(), ring.resultat[0].GetVida(),
			ring.resultat[1].GetNom(), ring.resultat[1].GetVida())
		elQuePica = elQueRep
	}

	guanyador := ring.resultat[1]
	perdedor := ring.resultat[0]
	if ring.resultat[1].EsKo() || ring.resultat[1].EstaEliminat() {
		guanyador = ring.resultat[0]
		perdedor = ring.resultat[1]
	}

	comentariLocutor := ""

	if perdedor.EstaEliminat() {
		comentariLocutor = perdedor.GetNom() + " està ELIMINAT per cops il·legals"
	} else {
		log.Printf("%s cau a terra!", perdedor.GetNom())
		if (guanyador.GetVida() - perdedor.GetVida()) > 5 {
			comentariLocutor = "Quina Pallissa!!"
		}
	}

	log.Printf("VICTÒRIA DE %s!!! %s", guanyador.GetNom(), comentariLocutor)

	return ring.resultat, nil
}

func contains(s []cops.LlocOnPicar, e cops.LlocOnPicar) bool {
	for _, a := range s {
		if a == e {
			return true
		}
	}
	return false
}
